flambe/nn/transformer.py

```diff
@@ -0,0 +1,532 @@
+
+        Parameters
+        ----------
+        tgt: torch.Tensor
+            The sequence to the decoder (required).
+        memory: torch.Tensor
+            The sequence from the last layer of the encoder (required).
+        tgt_mask: torch.Tensor, optional
+            The mask for the tgt sequence (optional).
+        memory_mask: torch.Tensor, optional
+            The mask for the memory sequence (optional).
+        tgt_key_padding_mask: torch.Tensor, optional
+            The mask for the tgt keys per batch (optional).
+        memory_key_padding_mask: torch.Tensor, optional
+            The mask for the memory keys per batch (optional).
+
+        Returns
+        -------
+        torch.Tensor
+
+        """
+        output = tgt
+
+        for i in range(self.num_layers):
+            output = self.layers[i](output,
+                                    memory,
+                                    tgt_mask=tgt_mask,
+                                    memory_mask=memory_mask,
+                                    tgt_key_padding_mask=tgt_key_padding_mask,
+                                    memory_key_padding_mask=memory_key_padding_mask)
+
+        return output
+
+    def _reset_parameters(self):
+        """Initiate parameters in the transformer model."""
+        for p in self.parameters():
+            if p.dim() > 1:
+                xavier_uniform_(p)
+
+
+class TransformerEncoderLayer(Module):
+    """TransformerEncoderLayer is made up of self-attn and feedforward.
+
+    This standard encoder layer is based on the paper "Attention Is
+    All You Need". Ashish Vaswani, Noam Shazeer, Niki Parmar,
+    Jakob Uszkoreit, Llion Jones, Aidan N Gomez, Lukasz Kaiser, and
+    Illia Polosukhin. 2017. Attention is all you need. In Advances in
+    Neural Information Processing Systems, pages 6000-6010. Users may
+    modify or implement in a different way during application.
+
+    """
+
+    def __init__(self,
+                 d_model: int,
+                 nhead: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 use_sru: bool = False) -> None:
+        """Initialize a TransformerEncoderLayer.
+
+        Parameters
+        ----------
+        d_model : int
+            The number of expected features in the input.
+        n_head : int
+            The number of heads in the multiheadattention models.
+        dim_feedforward : int, optional
+            The dimension of the feedforward network (default=2048).
+        dropout : float, optional
+            The dropout value (default=0.1).
+        use_sru: bool, optional
+            If true, replaces the FFN with an SRU.
+
+        """
+        super(TransformerEncoderLayer, self).__init__()
+
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+
+        if use_sru:
+            self.sru = SRU(d_model, dim_feedforward, num_layers=1, dropout=dropout)
```

Comment:
use `SRUCell` to avoid extra computation overhead.

Reply:
```
SRUCell(d_model, dim_feedforward, 
        dropout=dropout, 
        has_skip_term=False, 
        is_input_normalized=True)
```

Reply:
`bidirectional = True` or `False` should be configurable.

Reply:
Cool, will address this as part of your TransformerSRU suggestion